bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java

```diff
@@ -683,9 +744,73 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
 
 	@Override
 	public void setVisibleRegion(int start, int length) {
-		fWasProjectionEnabled= isProjectionMode();
-		disableProjection();
-		super.setVisibleRegion(start, length);
+		if (!isProjectionMode()) {
+			super.setVisibleRegion(start, length);
+			return;
+		}
+		IDocument document= getDocument();
+		if (document == null) {
+			return;
+		}
+		try {
+			// If the visible region changes, make sure collapsed regions outside of the old visible regions are expanded
+			// and collapse everything outside the new visible region
+			int end= computeEndOfVisibleRegion(start, length, document);
+			expandOutsideCurrentVisibleRegion(document);
+			collapseOutsideOfNewVisibleRegion(start, end, document);
+			fVisibleRegionDuringProjection= new Region(start, end - start);
+		} catch (BadLocationException e) {
+			String PLUGIN_ID= "org.eclipse.jface.text"; //$NON-NLS-1$
+			Bundle bundle= Platform.getBundle(PLUGIN_ID);
+			ILog log= ILog.of(bundle);
+			log.log(new Status(IStatus.WARNING, getClass(), IStatus.OK, e.getMessage(), e));
+		}
+		fVisibleRegionDuringProjection= new Region(start, length);
+	}
+
+	private void expandOutsideCurrentVisibleRegion(IDocument document) throws BadLocationException {
+		if (fVisibleRegionDuringProjection != null) {
+			expand(0, fVisibleRegionDuringProjection.getOffset(), false);
+			int oldEnd= fVisibleRegionDuringProjection.getOffset() + fVisibleRegionDuringProjection.getLength();
+			int length= document.getLength() - oldEnd;
+			if (length > 0) {
+				expand(oldEnd, length, false);
+			}
+		}
+	}
+
+	private void collapseOutsideOfNewVisibleRegion(int start, int end, IDocument document) throws BadLocationException {
+		int documentLength= document.getLength();
+		collapse(0, start, true);
+
+		int endInvisibleRegionLength= documentLength - end;
+		if (endInvisibleRegionLength > 0) {
+			collapse(end, endInvisibleRegionLength, true);
+		}
+	}
+
+	private static int computeEndOfVisibleRegion(int start, int length, IDocument document) throws BadLocationException {
+		int documentLength= document.getLength();
+		int end= start + length + 1;
+		// ensure that trailing whitespace is included
+		// In this case, the line break needs to be included as well
+		boolean visibleRegionEndsWithTrailingWhitespace= end < documentLength && isWhitespaceButNotNewline(document.getChar(end - 1));
+		while (end < documentLength && isWhitespaceButNotNewline(document.getChar(end))) {
```

Comment:
this iterates over every character in the document, but shouldn't iterate by lines and stop immediately current line checks if the character is not matching?